project/npda/tasks.py

```diff
@@ -0,0 +1,352 @@
+        elif isinstance(data, datetime) and data == datetime(1, 1, 1, 0, 0):
+            return None
+        else:
+            return data
+
+    def convert_iso_dates(obj):
+        """Recursively converts ISO 8601 date strings in a JSON-like object to datetime."""
+        if isinstance(obj, dict):
+            return {k: convert_iso_dates(v) for k, v in obj.items()}
+        elif isinstance(obj, list):
+            return [convert_iso_dates(i) for i in obj]
+        elif isinstance(obj, str):
+            try:
+                return (
+                    pd.to_datetime(obj) if "T" in obj else obj
+                )  # Check for ISO format
+            except ValueError:
+                return obj
+        return obj
+
+    def set_field_value_to_none_if_type_mismatch(form, visit_field):
+        """
+        Set the value of the field to None if the type is a mismatch
+        """
+        if form.fields[visit_field].widget.input_type == "select":
+            # expecting a key value
+            if field == "ethnicity":
+                # expecting a string
+                if type(visit_form.data[visit_field]) != str:
+                    setattr(visit_form.instance, field, None)
+            else:
+                # expecting an integer
+                if type(form.data[visit_field]) != int:
+                    setattr(form.instance, field, None)
+        elif form.fields[visit_field].widget.input_type == "date":
+            # expecting a date
+            if type(form.data[visit_field]) != datetime:
+                setattr(form.instance, field, None)
+
+    """
+    Main function
+    """
+
+    # Create a progress tracker
+    progress_tracker = ProgressTracker(self.request.id)
+    total_visits = len(patient_group_dict)
+
+    # import the models
+    Submission = apps.get_model("npda", "Submission")
+    PaediatricDiabetesUnit = apps.get_model("npda", "PaediatricDiabetesUnit")
+    Transfer = apps.get_model("npda", "Transfer")
+
+    print("running save patient in celery...")
+
+    # Gather all error messages indexed by row number and the field that caused them (__all__ if we don't know which one)
+    # dict[number, dict[str, list[str]]]
+    errors_to_return = collections.defaultdict(lambda: collections.defaultdict(list))
+    deserialized_patient_row = json.loads(patient_row_json)
+
+    patient_row = pd.Series(convert_iso_dates(deserialized_patient_row))
+    patient_group = pd.DataFrame(patient_group_dict)
+    patient_form = PatientForm(data=patient_dict)
+    pdu = PaediatricDiabetesUnit.objects.get(pz_code=pz_code)
+    submission = Submission.objects.get(id=submission_id)
+
+    # This protects the whole block of code from failing if there is an error and the database cannot be rolled back
+    # This is important because we are creating multiple objects in the database
+    # Hopefully it will never be needed, but it's good to have it just in case
+    with transaction.atomic():
```

Comment:
Wouldn't this remove the patient if one of their visits has a catastrophic error saving?